text/0000-RouteInfo-Metadata.md

```diff
@@ -0,0 +1,308 @@
+- Start Date: 2018-11-02
+- RFC PR: https://github.com/emberjs/rfcs/pull/398
+- Ember Issue: (leave this empty)
+
+# RouteInfo MetaData
+
+## Summary
+
+The RFC introduces the ability to associate application specific metadata with its corresponding `RouteInfo` object. This also adds a `metadata` field to `RouteInfo`, which will be the return value of `buildRouteInfoMetadata` for its corresponding `Route`.
+
+```js
+// app/route/profile.js
+import Route from '@ember/routing/route';
+import { inject as service } from '@ember/service';
+
+export default Route.extend({
+  user: inject('user'),
+  buildRouteInfoMetadata() {
+    return {
+      trackingKey: 'page_profile',
+      user: {
+        id: this.user.id,
+        type: this.user.type
+      }
+    }
+  }
+  // ...
+});
+```
+
+```js
+// app/services/analytics.js
+import Service, { inject } from '@ember/service';
+
+export default Service.extend({
+  router: inject('router'),
+  init() {
+    this._super(...arguments);
+    this.router.on('routeDidUpdate', (transition) => {
+      let { to, from } = transition;
+      let fromMeta = from.metadata;
+      let toMeta = to.metadata;
+      ga.sendEvent('pageView', {
+        from: fromMeta,
+        to: toMeta,
+        timestamp: Date.now(),
+      })
+    })
+  },
+  // ...
+});
+```
+
+## Motivation
+
+While the `RouteInfo` object is sufficient in providing developers metadata about the `Route` itself, it is not sufficient in layering on application specific metadata about the `Route`. This metadata could be anything from a more domain-specific name for a `Route`, e.g. `profile_page` vs `profile.index`, all the way to providing contextual data when the `Route` was visited.
+
+This metadata could be used for more pratical things like updating the `document.title`.
```

Comment:
I think `document.title` part in the Motivation section is misleading a bit.

While a proposed stateless `metadata` design allows you to cover titles for static routes. It doesn't seem possible to support `document.title` for routes with dynamic params. 

For example, let's say I have a route with a dynamic segment `/products/:id`. In this case I would expect to have a `product.title`(loaded in the model hook) included to the `document.title`.

Reply:
Any time we lean on the global `router.currentRoute` in an event handler like this, we are implicitly making timing guarantees about the relative order of events.

Maybe that is OK and unavoidable, but it's worth thinking about alternatives that make the data dependency more explicit:

```ts
this.router.on('routeDidUpdate', async (transition) => {
  let model = await transition.to.withAttributes();
  document.title = transition.to.title(model);
})
```

(in which I'm imagining a new `withAttributes(): Promise<RouteInfoWithAttributes>` method on `RouteInfo`.)

Reply:
There's a tradeoff here, because the simple thing introduces asynchrony, which then also invites a whole host of other issues.

Reply:
Wait, is there any reason `routeDidUpdate` can't receive a `transition` that already has `RouteInfoWithAttributes`? It seems like by the time we're doing didUpdate we have the info.